### ISSUE TITLE
Inference failure

### ISSUE THREAD

**Issue by charleskawczynski** [CliMA/ClimaCore.jl#1062]
Here is a MWE where we have inference failure, caught by Cthulhu (not JET, surprisingly), which results in allocations:

```julia
using Test

import ClimaCore
using ClimaCore: Geometry, Domains, Meshes, Topologies, Spaces, Fields

function toy_sphere(::Type{FT}) where {FT}
    helem = npoly = 2
    hdomain = Domains.SphereDomain(FT(1e7))
    hmesh = Meshes.EquiangularCubedSphere(hdomain, helem)
    htopology = Topologies.Topology2D(hmesh)
    quad = Spaces.Quadratures.GLL{npoly + 1}()
    hspace = Spaces.SpectralElementSpace2D(htopology, quad)
    vdomain = Domains.IntervalDomain(
        Geometry.ZPoint{FT}(zero(FT)),
        Geometry.ZPoint{FT}(FT(1e4));
        boundary_tags = (:bottom, :top),
    )
    vmesh = Meshes.IntervalMesh(vdomain, nelems = 4)
    vspace = Spaces.CenterFiniteDifferenceSpace(vmesh)
    center_space = Spaces.ExtrudedFiniteDifferenceSpace(hspace, vspace)
    face_space = Spaces.FaceExtrudedFiniteDifferenceSpace(center_space)
    return (;center_space,face_space)
end

function field_vec(center_space, face_space)
    Y = Fields.FieldVector(
        c = map(Fields.coordinate_field(center_space)) do coord
                FT = Spaces.undertype(center_space)
                (;
                    ρ = FT(0),
                    uₕ = Geometry.Covariant12Vector(FT(0), FT(0)),
                )
            end,
         f = map(Fields.coordinate_field(face_space)) do coord
                 FT = Spaces.undertype(face_space)
                 (;
                     w = Geometry.Covariant3Vector(FT(0)),
                 )
             end,
    )
    return Y
end

import ClimaTimeSteppers as CTS

function foo!(obj)
    (;u, b_exp, U) = obj
    @inbounds for i in 1:length(b_exp)
        @. U[i] = u
    end
    return nothing
end

function get_obj(nt)
    inds = ntuple(i -> i, length(nt.b_exp))
    U = CTS.SparseContainer(map(i -> similar(nt.u), collect(1:length(inds))), inds)
    return (;u = nt.u, b_exp = nt.b_exp, U)
end

using ClimaCore

(; center_space, face_space) = toy_sphere(Float64);
u = field_vec(center_space, face_space);

using StaticArrays
b_exp = @SVector [0.0, 1.20849664917601, -0.6443631706844692, 0.435866521508459]

obj = get_obj((;u, b_exp));

foo!(obj) # compile first

palloc = @allocated foo!(obj)
```

**Comment by charleskawczynski** [comment]
Scratch:

```julia
using Test

import ClimaCore
using ClimaCore: Geometry, Domains, Meshes, Topologies, Spaces, Fields

function toy_sphere(::Type{FT}) where {FT}
    helem = npoly = 2
    hdomain = Domains.SphereDomain(FT(1e7))
    hmesh = Meshes.EquiangularCubedSphere(hdomain, helem)
    htopology = Topologies.Topology2D(hmesh)
    quad = Spaces.Quadratures.GLL{npoly + 1}()
    hspace = Spaces.SpectralElementSpace2D(htopology, quad)
    vdomain = Domains.IntervalDomain(
        Geometry.ZPoint{FT}(zero(FT)),
        Geometry.ZPoint{FT}(FT(1e4));
        boundary_tags = (:bottom, :top),
    )
    vmesh = Meshes.IntervalMesh(vdomain, nelems = 4)
    vspace = Spaces.CenterFiniteDifferenceSpace(vmesh)
    center_space = Spaces.ExtrudedFiniteDifferenceSpace(hspace, vspace)
    face_space = Spaces.FaceExtrudedFiniteDifferenceSpace(center_space)
    return (;center_space,face_space)
end

function field_vec(center_space, face_space)
    Y = Fields.FieldVector(
        c = map(Fields.coordinate_field(center_space)) do coord
                FT = Spaces.undertype(center_space)
                (;
                    ρ = FT(0),
                    uₕ = Geometry.Covariant12Vector(FT(0), FT(0)),
                )
            end,
        # f = map(Fields.coordinate_field(face_space)) do coord
        #         FT = Spaces.undertype(face_space)
        #         (;
        #             w = Geometry.Covariant3Vector(FT(0)),
        #         )
        #     end,
    )
    return Y
end

import ClimaTimeSteppers as CTS

function foo!(obj)
    (;u, b_exp, U) = obj
    @inbounds for i in 1:length(b_exp)
        @. U[i] = u
    end
    return nothing
end

function get_obj(nt)
    inds = ntuple(i -> i, length(nt.b_exp))
    U = CTS.SparseContainer(map(i -> similar(nt.u), collect(1:length(inds))), inds)
    return (;u = nt.u, b_exp = nt.b_exp, U)
end

using ClimaCore

(; center_space, face_space) = toy_sphere(Float64);
u = field_vec(center_space, face_space);

using StaticArrays
b_exp = @SVector [0.0, 1.20849664917601, -0.6443631706844692, 0.435866521508459]

obj = get_obj((;u, b_exp));

foo!(obj) # compile first

palloc = @allocated foo!(obj)
@show palloc

using BenchmarkTools;
@benchmark foo!(obj)

import Cthulhu
Cthulhu.@descend foo!(obj) # shows inference failures in materialize!

import JET
JET.@test_opt foo!(obj) # passes
```

**Comment by charleskawczynski** [comment]
Closed by #1065
